File: AnalysisCode/OPGEE_rows_func.m
function [tranche_OPGEE] = OPGEE_rows_func(Replicate, OPGEE_bin, LU_type, flare_tab)


% OPGEE binning func

      
% For dry gas bins
tranche_data = OPGEE_bin.gasdry;
tranche_OPGEE = [];
% if Basin_Select == 0
%  C1_frac = 83.2; 
% end

for i = 1: length(tranche_data(:,1))
    for j = 1:3
        wells = tranche_data(i,1) * LU_type(j);
        wells = round(wells);
        prod_mscf = tranche_data(i,3) * LU_type(j);
        prod_bbl = tranche_data(i,4) * LU_type(j);
        if Replicate  ~= 1
            flare_frac = flare_tab.gasdry(i,4) * LU_type(j);
        else
            flare_frac = 0;
        end
        
        if prod_bbl < 0.4
            prod_bbl = 0.4;
        end
        
        prod_GOR = (prod_mscf*1000)/prod_bbl;
        
        vec = [prod_bbl wells (tranche_data(i,16)*100) prod_GOR j flare_frac tranche_data(i,5:15)];
        tranche_OPGEE = [tranche_OPGEE; vec];
    end    
end

% For gas wells with oil
tranche_data = OPGEE_bin.gasassoc;
% if Basin_Select == 0
%  C1_frac = 83.2; 
% end

for i = 1: length(tranche_data(:,1))
    for j = 1:3
        wells = tranche_data(i,1) * LU_type(j);
        wells = round(wells);
        prod_mscf = tranche_data(i,3) * LU_type(j);
        prod_bbl = tranche_data(i,4) * LU_type(j);
        if Replicate  ~= 1
            flare_frac = flare_tab.gasassoc(i,4) * LU_type(j);
        else
            flare_frac = 0;
        end
        
    
        if prod_bbl < 0.4
            prod_bbl = 0.4;
        end
        
        prod_GOR = (prod_mscf*1000)/prod_bbl;
           
        vec = [prod_bbl wells (tranche_data(i,16)*100) prod_GOR j flare_frac tranche_data(i,5:15)];
        tranche_OPGEE = [tranche_OPGEE; vec];
    end    
end

% For oil wells with gas
tranche_data = OPGEE_bin.oilwgas;
% if Basin_Select == 0
%  C1_frac = 68.3; 
% end

for i = 1: length(tranche_data(:,1))
    wells = tranche_data(i,1);
    wells = round(wells);
    prod_mscf = tranche_data(i,3);
    prod_bbl = tranche_data(i,4);
    if Replicate  ~= 1
        flare_frac = flare_tab.oilwgas(i,4);
    else
        flare_frac = 0;
    end

    if prod_bbl < 0.4
        prod_bbl = 0.4;
    end
    
    prod_GOR = (prod_mscf*1000)/prod_bbl;
            
    vec = [prod_bbl wells (tranche_data(i,16)*100) prod_GOR 3 flare_frac tranche_data(i,5:15)];
    tranche_OPGEE = [tranche_OPGEE; vec];   
end

% For oil wells
tranche_data = OPGEE_bin.oil;
% if Basin_Select == 0
%  C1_frac = 68.3; 
% end

for i = 1: length(tranche_data(:,1))
    wells = tranche_data(i,1);
    wells = round(wells);
    prod_mscf = tranche_data(i,3);
    prod_bbl = tranche_data(i,4);
    if Replicate  ~= 1
        flare_frac = flare_tab.oil(i,4);
    else
        flare_frac = 0;
    end
    
    if prod_bbl < 0.4
        prod_bbl = 0.4;
    end
    
    prod_GOR = (prod_mscf*1000)/prod_bbl;
        
    vec = [prod_bbl wells (tranche_data(i,16)*100) prod_GOR 3 flare_frac tranche_data(i,5:15)];
    tranche_OPGEE = [tranche_OPGEE; vec];   
end

% csvwrite(char(strcat(Basin_Index(1),'_OPGEE_trch.csv')),tranche_OPGEE)

end

File: AnalysisCode_python/OPGEE_rows_func.py
import numpy as np

def OPGEE_rows_func(Replicate, OPGEE_bin, LU_type, flare_tab):
    tranche_OPGEE = []

    # For dry gas bins
    tranche_data = OPGEE_bin['gasdry']

    for i in range(len(tranche_data)):
        for j in range(3):
            wells = round(tranche_data[i, 0] * LU_type[j])
            prod_mscf = tranche_data[i, 2] * LU_type[j]
            prod_bbl = tranche_data[i, 3] * LU_type[j]
            flare_frac = flare_tab['gasdry'][i, 3] * LU_type[j] if Replicate != 1 else 0

            if prod_bbl < 0.4:
                prod_bbl = 0.4

            prod_GOR = (prod_mscf * 1000) / prod_bbl
            vec = [prod_bbl, wells, tranche_data[i, 15] * 100, prod_GOR, j + 1, flare_frac] + list(
                tranche_data[i, 4:15])  # Adjust indexing to capture the same columns as in MATLAB
            tranche_OPGEE.append(vec)

    # For gas wells with oil
    tranche_data = OPGEE_bin['gasassoc']

    for i in range(len(tranche_data)):
        for j in range(3):
            wells = round(tranche_data[i, 0] * LU_type[j])
            prod_mscf = tranche_data[i, 2] * LU_type[j]
            prod_bbl = tranche_data[i, 3] * LU_type[j]
            flare_frac = flare_tab['gasassoc'][i, 3] * LU_type[j] if Replicate != 1 else 0

            if prod_bbl < 0.4:
                prod_bbl = 0.4

            prod_GOR = (prod_mscf * 1000) / prod_bbl
            vec = [prod_bbl, wells, tranche_data[i, 15] * 100, prod_GOR, j + 1, flare_frac] + list(
                tranche_data[i, 4:15])  # Adjust indexing to capture the same columns as in MATLAB
            tranche_OPGEE.append(vec)

    # For oil wells with gas
    tranche_data = OPGEE_bin['oilwgas']

    for i in range(len(tranche_data)):
        wells = round(tranche_data[i, 0])
        prod_mscf = tranche_data[i, 2]
        prod_bbl = tranche_data[i, 3]
        flare_frac = flare_tab['oilwgas'][i, 3] if Replicate != 1 else 0

        if prod_bbl < 0.4:
            prod_bbl = 0.4

        prod_GOR = (prod_mscf * 1000) / prod_bbl
        vec = [prod_bbl, wells, tranche_data[i, 15] * 100, prod_GOR, 3, flare_frac] + list(tranche_data[i, 4:15])  # Adjust indexing
        tranche_OPGEE.append(vec)

    # For oil wells
    tranche_data = OPGEE_bin['oil']

    for i in range(len(tranche_data)):
        wells = round(tranche_data[i, 0])
        prod_mscf = tranche_data[i, 2]
        prod_bbl = tranche_data[i, 3]
        flare_frac = flare_tab['oil'][i, 3] if Replicate != 1 else 0

        if prod_bbl < 0.4:
            prod_bbl = 0.4

        prod_GOR = (prod_mscf * 1000) / prod_bbl
        vec = [prod_bbl, wells, tranche_data[i, 15] * 100, prod_GOR, 3, flare_frac] + list(tranche_data[i, 4:15])  # Adjust indexing
        tranche_OPGEE.append(vec)

    return np.array(tranche_OPGEE)
